demo-circuit/src/blaze_csw/mod.rs

```diff
@@ -0,0 +1,326 @@
+            .inverse()
+            .expect("Should be able to compute inverse of (1 - y_te) ");
+
+    let sw_x_coord = (one_plus_te_y_coord_over_one_minus_te_y_coord + a_over_three) * b_inv;
+    let sw_y_coord = b_inv * one_plus_te_y_coord_over_one_minus_te_y_coord * te_x_coord_inv;
+
+    let sw_point = SimulatedSWGroup::new(sw_x_coord, sw_y_coord, false);
+
+    debug_assert!(sw_point.group_membership_test());
+
+    sw_point
+}
+
+/// Convert a sc pk, expressed in TE form, to SW, to be able to use it inside the blaze csw circuit.
+pub fn convert_te_pk_to_sw_pk(
+    mut te_pk_bytes: [u8; SC_PUBLIC_KEY_LENGTH],
+) -> Result<[u8; SC_PUBLIC_KEY_LENGTH], Error> {
+    // First, let's reconstruct the TE point corresponding to te_pk_bytes
+
+    // Fetch the sign of the x coordinate
+    let te_pk_x_sign = !((te_pk_bytes[SC_PUBLIC_KEY_LENGTH - 1] & (1 << 7)) == 0u8);
+
+    // Mask away the sign byte
+    te_pk_bytes[SC_PUBLIC_KEY_LENGTH - 1] &= 0x7F;
+
+    // Deserialize the y coordinate
+    let te_pk_y = deserialize_from_buffer::<SimulatedFieldElement>(&te_pk_bytes, None, None)?;
+
+    // Reconstruct the x coordinate from the y coordinate and the sign
+    let te_pk_x = {
+        let numerator = te_pk_y.square() - SimulatedFieldElement::one();
+        let denominator = (te_pk_y.square() * SimulatedCurveParameters::COEFF_D)
+            - &<SimulatedCurveParameters as TEModelParameters>::COEFF_A;
+        let x2 = denominator.inverse().map(|denom| denom * numerator);
+        x2.and_then(|x2| x2.sqrt()).map(|x| {
+            let negx = -x;
+            if x.is_odd() ^ te_pk_x_sign {
+                negx
+            } else {
+                x
+            }
+        })
+    }
+    .ok_or_else(|| Error::from("Invalid pk. Unable to reconstruct x coordinate."))?;
+
+    // Reconstruct the TE point and check that it's on curve
+    let te_pk = SimulatedTEGroup::new(te_pk_x, te_pk_y);
+    if te_pk.is_zero() || !te_pk.group_membership_test() {
+        return Err(Error::from("Error: invalid pk"));
+    }
+
+    // Now, convert the TE point to a SW one.
+    // Conversion formulas:
+    // https://en.wikipedia.org/wiki/Montgomery_curve#Equivalence_with_twisted_Edwards_curves
+    // https://en.wikipedia.org/wiki/Montgomery_curve#Equivalence_with_Weierstrass_curves
+    let sw_pk = convert_te_point_to_sw_point(te_pk);
+
+    // Store the sign (last bit) of the X coordinate
+    // The value is left-shifted to be used later in an OR operation
+    let x_sign = if sw_pk.x.is_odd() { 1 << 7 } else { 0u8 };
+
+    // Extract the public key bytes as Y coordinate
+    let y_coordinate = sw_pk.y;
+    let mut pk_bytes = serialize_to_buffer(&y_coordinate, None).unwrap();
+
+    // Use the last (null) bit of the public key to store the sign of the X coordinate
+    // Before this operation, the last bit of the public key (Y coordinate) is always 0 due to the field modulus
+    let len = pk_bytes.len();
+    pk_bytes[len - 1] |= x_sign;
```

Comment:
This is non-injective serialization: Weierstrass curves can have up to three points  for a given y-value.
I propose to use serialize via x-coordinate, plus the parity bit for y.

Reply:
Yes you're right, it's not used but I changed it